### PR TITLE
`_vispy` module cleanup

### PR DESCRIPTION
**Pull request by brisvag**
# Description
As discussed in the architecture meetings, the `_vispy` module is a bit messy and there are quite a few inconsistencies in how vispy visuals and layers are defined. In an effort to streamline this part of the codebase, I started thinking on how to refactor some of this code into a clearly designed module.

~Since this takes a bit of time, I'm putting this very tentative PR up in order to gather feedback on stuff like file/directory names before I start the search:replace hunt :P~ I ended up doing it anyways. It should mostly be working now.

Notable changes:
- Separated the concepts of `Visual`, `Layer` and `Overlay` (Previously also called Visual, despite being closer to a `VispyLayer` than a vispy `Visual`)
-  removed a lot of redundant `vispy`s in directory and file names (but left in class names)
- added a mixin for clipping planes, to remove the duplicated code.

For the future (probably not this PR):
- It would probably be nice to add a superclass (or mixin) for `Overlay`s since they share some functionality.
- It would be great to have an automated way to connect all those events...

This is how the module looks like now:
```
.
├── camera.py
├── canvas.py
├── experimental
│   ├── __init__.py
│   ├── _tests
│   ├── texture_atlas.py
│   ├── tiled_image_visual.py
│   ├── tile_grid.py
│   ├── tile_set.py
│   └── vispy_tiled_image_layer.py
├── filters
│   ├── __init__.py
│   └── tracks.py
├── __init__.py
├── layers
│   ├── base.py
│   ├── image.py
│   ├── __init__.py
│   ├── points.py
│   ├── shapes.py
│   ├── surface.py
│   ├── tracks.py
│   └── vectors.py
├── overlays
│   ├── axes.py
│   ├── __init__.py
│   ├── scale_bar.py
│   └── text.py
├── _tests
│   ├── test_image_rendering.py
│   ├── test_utils.py
│   ├── test_vispy_big_images.py
│   ├── test_vispy_calls.py
│   ├── test_vispy_camera.py
│   ├── test_vispy_multiscale.py
│   ├── test_vispy_points_layer.py
│   ├── test_vispy_scale_bar_visual.py
│   └── test_vispy_text_visual.py
├── utils
│   ├── gl.py
│   ├── __init__.py
│   ├── quaternion.py
│   ├── text.py
│   └── visual.py
├── vendored
│   ├── filters
│   ├── gloo
│   ├── image.py
│   ├── __init__.py
│   ├── mesh.py
│   ├── _scalable_textures.py
│   ├── volume.py
│   └── volume.py.orig
└── visuals
    ├── clipping_planes_mixin.py
    ├── image.py
    ├── __init__.py
    ├── markers.py
    ├── mesh.py
    ├── points.py
    ├── shapes.py
    ├── surface.py
    ├── tracks.py
    ├── vectors.py
    └── volume.py
```
This is what it looks like on master:
```
.
├── experimental
│   ├── __init__.py
│   ├── _tests
│   ├── texture_atlas.py
│   ├── tiled_image_visual.py
│   ├── tile_grid.py
│   ├── tile_set.py
│   └── vispy_tiled_image_layer.py
├── filters
├── image.py
├── __init__.py
├── layers
├── markers.py
├── mesh.py
├── overlays
├── quaternion.py
├── _tests
│   ├── test_image_rendering.py
│   ├── test_utils.py
│   ├── test_vispy_big_images.py
│   ├── test_vispy_calls.py
│   ├── test_vispy_camera.py
│   ├── test_vispy_multiscale.py
│   ├── test_vispy_points_layer.py
│   ├── test_vispy_scale_bar_visual.py
│   └── test_vispy_text_visual.py
├── _text_utils.py
├── utils
├── utils_gl.py
├── utils.py
├── vendored
│   ├── filters
│   ├── gloo
│   ├── image.py
│   ├── __init__.py
│   ├── mesh.py
│   ├── _scalable_textures.py
│   ├── volume.py
│   └── volume.py.orig
├── vispy_axes_visual.py
├── vispy_base_layer.py
├── vispy_camera.py
├── vispy_canvas.py
├── vispy_image_layer.py
├── vispy_points_layer.py
├── vispy_points_visual.py
├── vispy_scale_bar_visual.py
├── vispy_shapes_layer.py
├── vispy_shapes_visual.py
├── vispy_surface_layer.py
├── vispy_surface_visual.py
├── vispy_text_visual.py
├── vispy_tracks_layer.py
├── _vispy_tracks_shader.py
├── vispy_tracks_visual.py
├── vispy_vectors_layer.py
├── visuals
└── volume.py
```

## Type of change

- [x] Refactor

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by brisvag** [comment]
@quantumjot, @jni suggested to ask you about this: what is the purpose of the two different `TrackShader` objects for `tracks` and `graph`. They look to me like they could just be the same object, since they always get updated by the same events syncronously. Is that right?

File: napari/_vispy/visuals/points.py
Comment (andy-sweet) [inline napari/_vispy/visuals/points.py]: I guess these were moved from elsewhere, but I would remove this implementation comment and absorb any unique information in the docstring above. Currently, this is missing the `TextVisual` and mostly contains the same information.

File: napari/_vispy/visuals/shapes.py
Comment (andy-sweet) [inline napari/_vispy/visuals/shapes.py]: Absorb into docstring.

File: napari/_vispy/visuals/clipping_planes_mixin.py
Comment (andy-sweet) [inline napari/_vispy/visuals/clipping_planes_mixin.py]: I'm still getting used to (and warming up to) mixins, but it seems a bit odd that this class knows something about the classes that use it - in this case that either are a vispy visual or have some in `_subvisuals`.

I have no idea if this standard for mixins, but one thing that might work well is to define an abstract/not-implemented method called `get_visuals()` in this mixin that must be implemented by the class that uses it. That adds a small amount of code to each of the classes that use this, but it also enforces the contract that you expect here earlier/more explicitly than this does.

As I said, I'm still new to mixins, and I think it's unlikely that `ClippingPlanesMixin` will be used with other things, so feel free to ignore this.

File: napari/_vispy/visuals/surface.py
Comment (andy-sweet) [inline napari/_vispy/visuals/surface.py:5]: Could `VectorsVisual` also be defined this way, or is there a reason that a vectors layer could/should not be clipped?

**Review by andy-sweet — COMMENTED** [review]
Overall, this looks like a good set of changes, though I'm not too familiar with this part of the code, so I can't be too helpful (yet).

File: napari/_vispy/visuals/points.py
Comment (brisvag) [reply napari/_vispy/visuals/points.py]: Yup, thanks for pointing out, this got lost in the refactoring!

File: napari/_vispy/visuals/clipping_planes_mixin.py
Comment (brisvag) [reply napari/_vispy/visuals/clipping_planes_mixin.py]: I might be using the terminology completely wrong, so please correct me if "mixin" is not the right therm here :)
The reason this exists is mainly because `Volume` behaves differently from all the other visuals (no use of the filter, despite having the same high level api of `Volume.clipping_planes = value`), so plain old inheritance was not the right way. do you have a better suggestion?

File: napari/_vispy/visuals/surface.py
Comment (brisvag) [reply napari/_vispy/visuals/surface.py:5]: Oops. This is actually the visual that should be used by the vectors layer :)

**Review by tlambert03 — APPROVED** [review]
I like it!  looks much cleaner to me and I think all the wording changes make sense

File: napari/_vispy/visuals/clipping_planes_mixin.py
Comment (andy-sweet) [reply napari/_vispy/visuals/clipping_planes_mixin.py]: I'm not familiar enough to give strong guidance on whether mixin is the right term here - but from I what I understand of mixins and your intent, I think it is because as you said you want to some behavior to both Mesh and Compound visuals, but not Volume visuals.

Your recent update is an improvement because now the assumption about the class that uses the mixin is simpler - that it implements an `attach` method.

If it's well understood that `ClippingPlanesMixin` will always be used by a class that implements or inherits an `attach` method, then this is good enough.

If you wanted to protect misuse of `ClippingPlanesMixin` (i.e. by classes that don't have an `attach` method), then you could move the call to attach in each of the Visuals' (e.g. `ShapesVisual.__init__`) - it's slightly more code (one line per Visual class), but it's also easier to understand because that line of code should be close to the declaration of the class where we can see it inherits from `Compound` or `Mesh` (i.e. the implementors of `attach`).

Comment (tlambert03) [reply napari/_vispy/visuals/clipping_planes_mixin.py]: if you like, you can indicate the intent of the Mixin (and provide some IDE support if someone uses it wrong) with:
```python
from typing import List, Optional
from typing_extensions import Protocol

class _PVisual(Protocol):
    _subvisuals: Optional[List['_PVisual']]
    def attach(self, filt: Filter, view=None ): ...

class ClippingPlanesMixin:
    """
    Mixin class that attaches clipping planes filters to the (sub)visuals
    and provides property getter and setter
    """

    def __init__(self: _PVisual, *args, **kwargs):
```

**Comment by brisvag** [comment]
I'm having a failure in py37 min-req only, where it can't import `Filter` from vispy. How do I get around that?

**Comment by tlambert03** [comment]
Did filter not exist in 0.6.4? Or did it move?

Since this is only a type hint, you can hide the import behind `if TYPE_CHECKING`, and then 'Filter' (as a string) for the hint.

But do we not actually use Filter anywhere?  If we do, you could bump our minimum vispy requirement?

**Comment by brisvag** [comment]
We were already using `Filter`, at least since 18f97f08f. Not sure why it would fail now. Maybe it's specifically a typing issue?

**Comment by brisvag** [comment]
Right, I see now. In 0.6.4 it wasn't imported at `filters` module level :)

**Comment by brisvag** [comment]
All CI issues seem solved. I think this can be merged!

**Comment by tlambert03** [comment]
looks great to me @brisvag.  moved will merge in a day or two if no one has any objectiions

**Review by jni — APPROVED** [review]
I'm happy with this (though @brisvag there is a merge conflict now). I'll say one thing which is that all of the files were originally prefixed with `vispy_` because many/most IDEs only put the basename in the tab, so it was annoying to not know which `image.py` I was looking at. But I think it's a minor concern, this makes more sense in the end to a newcomer.

**Review by sofroniewn — APPROVED** [review]
This looks great, i like the new folders, we've now got merge conflicts, but once they are addressed we can merge

**Comment by goanpeca** [comment]
This one looks ready to go. Thanks a lot @brisvag !

**Pull request merged**
